### DISCUSSION TITLE
Import Error Caused by Error of and Error file multiple levels down

### DISCUSSION THREAD

**Discussion by DS256** [micropython/micropython#11642] [RP2040 / Pico] [answered]
This is the 2nd time I've run into this issue so thought I'd share. Before I detail the issue, you should understan a couple of things I've done with my uPython programming on my PICO W.

- My BOOT.PY is a simple program to call 'boot_x()' in BOOT_X.MPY. I do this for a couple of reason. First is that in managing my programs, it provides a unique file name. Second is that by using MPY it reduces the size of the program. BOOT_X.MPY is responsible for internet connectivity and for getting and setting the RTC.
- I also have a similar MAIN.PY that calls 'main_x()' from MAIN_X,MPY files.
- I use parameter files to avoid having to make program changes, especially when using MPY's. This also applies to shared functions which I import to different PICO W mPython programs. These contain variable assignments only. They are imported as 'import boot_x_params as params'. The variables are then reference as 'params.y' where y is the variable in the file.

Here is the problem. I was testing a change and suddenly I was not able to import 'boot_x()' from BOOT_X.MPY. I then tried to import the whole BOOT_X which did not work.

The actual errors reported by uPython from running BOOT.PY

- Using '_from boot_x import boot_x_' = '**ImportError: can't import name boot_x**'
- Using '_import boot_x' when 'boot_x()_' was called = "**TypeError: 'module' object is not callable**'

The problem? 'boot_x()' was importing a revised 'set_rtc()' from SET_RTC.MPY. This function imports 'SET_RTC_PARAMS.PY'. I had made an incorrect change in this file. Here is the error file.
```
# Datetime Server used in set_rtc()
world_datetime_server="http://worldtimeapi.org/api/timezone/America/Toronto"
timezoneapi_server="https://timezoneapi.io/api/ip/?token=xxx&only=datetime"
params.retry_time=90 # Time to wait between failed attempts at getting time
```

I had cut and pasted the reference 'params.retry_time' in SET_RTC.PY to SET_RTC_PARAMS.PY' without removing the 'params.' prefix. It was the 'params.retry_time' that was causing the problem running BOOT.PY. Changing 'params.retry_time' to 'retry_time' solved the problem.

It took me awhile to ask myself 'what had changed?' when the problem reported was no where need the change was 3 import levels away.

Hope this helps someone in the future.

**Comment by jimmo** [comment]
@DS256 Can you make a simple example to demonstrate this and post the full code here. I tried to replicate it here, but got the expected result.

I have the following

boot.py
```py
from boot_x import boot_x
boot_x()
```

boot_x.py
```py
import foo

def boot_x():
    print("hello")
```

foo.py
```py
1/0
```

boot_x.py and foo.py are compiled to .mpy before being copied to the device. When I do a reset, I see

```
MPY: soft reboot
Traceback (most recent call last):
  File "boot.py", line 1, in <module>
  File "boot_x.py", line 1, in <module>
  File "foo.py", line 1, in <module>
ZeroDivisionError: divide by zero
```

If however I now do the following at the REPL:
```
>>> from boot_x import boot_x
```
I see `ImportError: can't import name boot_x`. This is because boot_x is already imported (but because it failed early, it does not contain the boot_x function). This is actually a bug (a failed import should be re-tried), will be fixed by #11421 which will be merged after we merge #11456

But I don't think it sounds like you're doing this at the REPL?

**Comment by DS256** [comment]
Hi @jimmo . Here's a demonstration
The version uP I'm using is "MicroPython 699477d on 2023-01-16; Raspberry Pi Pico W with RP2040" with Thonny 4.0.0.1.
Keep in mind, the purpose of my note is to show that a programming error far from where uPython reports an error is possible.

Here are my programs.
```
# BOOT.PY - Initial BOOT

from boot_x import boot_x
boot_x()
```
```
# BOOT_X.PY - Mainline Program

import boot_x_params as params

def boot_x():
    print("Hello from boot_x")
    print(params.a)
    print(params.b)
```
```
# BOOT_X_PARAMS.PY - Parameter File for BOOT_X.P
a=42
params.b="Oh no Mr. Bill" # This is the error that causes the problem
```
No need to compile to get the error.
```
>>> %Run -c $EDITOR_CONTENT
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ImportError: can't import name boot_x
>>> 
```
Changing 'params.b' to 'b' in 'boot_x_params.py' produces:
```
>>> %Run -c $EDITOR_CONTENT
Hello from boot_x
42
Oh no Mr. Bill
>>> 
```
Hope this clarfies

**Reply by jimmo** [reply]
@DS256 OK yeah you're seeing exactly what I described.

Thonny is complicating the issue here because when you click "Run" in thonny it's not actually the same as if boot.py is executing normally at boot time, rather it's as if you typed the lines of boot.py into the REPL.

The actual error happened at boot time, but you never saw that (not sure why, maybe Thonny wasn't connected?). Now what you see is that boot_x is partially imported, and therefore it's an empty module.

**Reply by DS256** [reply]
@jimmo then why:

- Does the problem go away when I change 'params.b' to 'b'?
- Why doesn't uP report the error in BOOT_X_PARAMS.PY.?

**Reply by jimmo** [reply, marked as answer]
> Does the problem go away when I change 'params.b' to 'b'?

Because boot.py will now successfully import boot_x at actual boot time.

> Why doesn't uP report the error in BOOT_X_PARAMS.PY.?

It does (see my example). I suspect you're not seeing it in Thonny for some reason.

Just for completeness, here's your exact example using the official tools. First shows the contents of the three files on the device, and the error message being shown at boot time.

```
$ mpr cat :boot.py
from boot_x import boot_x
boot_x()

$ mpr cat :boot_x.py
import boot_x_params as params

def boot_x():
    print("Hello from boot_x")
    print(params.a)
    print(params.b)

$ mpr cat :boot_x_params.py
a=42
params.b="Oh no Mr. Bill" # This is the error that causes the problem

$ mpr
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell

MPY: soft reboot
Traceback (most recent call last):
  File "boot.py", line 1, in <module>
  File "boot_x.py", line 1, in <module>
  File "boot_x_params.py", line 2, in <module>
NameError: name 'params' isn't defined
MicroPython v1.20.0-119-g59b5f3462-dirty on 2023-05-29; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> 
```

If I now try and import from boot_x at the REPL:

```
>>> from boot_x import boot_x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: can't import name boot_x
```

You see that it cannot get the function boot_x from the module because boot_x is an empty module (due to failing to import).

Like I said above, this is a longstanding missing feature (a failed import should lead to a re-try) that will be fixed soon, but the error message is definitely being displayed at boot time.

**Reply by DS256** [reply]
Thanks @jimmo. I was not aware of [MPR](https://pypi.org/project/mpr/). I'll add this to my test procedure.
I'll also follow up on the Thonny forum to make them aware. I didn't think the IDE could have so much influence but then I'm relatively new my uPython and the RP2040.

**Reply by jimmo** [reply]
`mpr` is just my local alias for `mpremote`, going back to when mpremote was first developed and that's what it was originally called. (I need to remember that when showing examples)

https://docs.micropython.org/en/latest/reference/mpremote.html

The `mpr` you linked to is actually a different thing, but based somehow on mpremote.

**Reply by DS256** [reply]
Thanks for the correction. I'll try mpremote. I had been using rshell which I thought was an official utility.
